WIP - Kubeflow Graduation DD (#4)
kfaseela/toc: projects/kubeflow/kubeflow-graduation-dd.md

- The Kubeflow maintainers and community were highly responsive throughout this review, with PRs merged, issues filed and assigned, and feedback addressed quickly.
- All graduation subprojects hold at least an OpenSSF Best Practices Passing badge; Kubeflow Trainer and Kubeflow Katib have achieved Gold.
- The Kubeflow Ecosystem page and process were established during this review to clarify the boundary between core subprojects and ecosystem partners such as KServe, Feast, and Elyra.
- Kubeflow shows broad adoption across research, financial services, logistics, and platform teams. Four adopter interviews were conducted across the graduation scope.
- Contributor health remains strong, with active contributors from a wide range of organizations and positive YoY growth.

DD Findings Resolved (Blocking/Required):

- **[kubeflow/community#992](https://github.com/kubeflow/community/issues/992) - Third-party security audit findings**: Ada Logics audit is published; findings are resolved or tracked. KServe findings are out of scope (independent CNCF project).

> 

Reviewer: ```suggestion
```

- [X] **Review and acknowledgement of expectations for sandbox projects and requirements for moving forward through the CNCF Maturity levels.**
  - [X] Acknowledged in the graduation application body (September 2025) per CNCF application requirements.

Completion of this due diligence document, resolution of concerns raised, and presentation for public comment satisfies the Due Diligence Review criterion.

- [X] **Additional documentation as appropriate for project type, e.g.: installation documentation, end user documentation, reference implementation and/or code samples.**

> All 6 graduation subprojects have dedicated documentation pages on kubeflow.org with installation guides, user guides, operator guides, and reference documentation. The [kubeflow.org docs site](https://www.kubeflow.org/docs/) is actively maintained. The [General Technical Review](https://github.com/kubeflow/community/blob/master/KUBEFLOW-GENERAL-TECHNICAL-REVIEW.md) (GTR) provides an extensive technical reference. Kubeflow Notebooks has an architectural dependency on Kubeflow Central Dashboard and cannot be deployed standalone - this is a known design characteristic, not a documentation gap.

Reviewer: ```suggestion
All 6 graduation subprojects have dedicated documentation pages on kubeflow.org with installation guides, user guides, operator guides, and reference documentation. The [kubeflow.org docs site](https://www.kubeflow.org/docs/) is actively maintained. The [General Technical Review](https://github.com/kubeflow/community/blob/master/KUBEFLOW-GENERAL-TECHNICAL-REVIEW.md) (GTR) provides an extensive technical reference. Kubeflow Notebooks has an architectural dependency on Kubeflow Central Dashboard and cannot yet be deployed standalone - this is however a part of their roadmap.
```

- [X] **Demonstrate usage of the maintainer lifecycle with outcomes, either through the addition or replacement of maintainers as project events have required.**

Addition examples (all verified merged 2025-2026):
- [Add @mprahl and @zazulam as KFP maintainers (pipelines#12059)](https://github.com/kubeflow/pipelines/pull/12059) - Added @mprahl + @zazulam as KFP maintainers with community announcement
- [Add @Electronic-Waste as approver, @astefanutti as reviewer (trainer#2659)](https://github.com/kubeflow/trainer/pull/2659) - Added with explicit criteria verification against the membership page
- [Add @Al-Pragliola as approver (hub#1153)](https://github.com/kubeflow/hub/pull/1153) - Added with full checklist verification and linked internal-acls PR

Removal example:
- [@zw0610 moves to emeritus_approvers (trainer#2343)](https://github.com/kubeflow/trainer/pull/2343) - @zw0610 self-removed to `emeritus_approvers`

- [X] **Project maintainers from at least 2 organizations that demonstrates survivability.**

> Maintainer org diversity is broad across all graduation subprojects, with approvers spanning a wide range of organizations ([LFX Insights](https://insights.lfx.dev/foundation/cncf/overview/github?project=kubeflow)). Org affiliation for governance-level maintainers is documented in the KSC and KOC member lists linked from [MAINTAINERS.md](https://github.com/kubeflow/community/blob/master/MAINTAINERS.md), and for WG leads/chairs in [wgs.yaml](https://github.com/kubeflow/community/blob/master/wgs.yaml). No subproject is controlled by a single organization. WG Notebooks leadership was expanded via [Update WG Notebooks docs (community#983)](https://github.com/kubeflow/community/pull/983), resolving a prior single-org concern.

Reviewer: ```suggestion
Maintainer org diversity is broad across all graduation subprojects, with approvers spanning a wide range of organizations ([LFX Insights](https://insights.lfx.dev/foundation/cncf/overview/github?project=kubeflow)). Org affiliation for governance-level maintainers is documented in the KSC and KOC member lists linked from [MAINTAINERS.md](https://github.com/kubeflow/community/blob/master/MAINTAINERS.md), and for WG leads/chairs in [wgs.yaml](https://github.com/kubeflow/community/blob/master/wgs.yaml). No subproject is controlled by a single organization. WG Notebooks leadership was updated in [Update WG Notebooks docs (community#983)](https://github.com/kubeflow/community/pull/983), to replace an inactive lead and ensure organizational diversity.
```

Community-wide changes use the [KEP process](https://github.com/kubeflow/community/tree/master/proposals) (proposal -> community review -> implementation). Subproject-specific proposals are managed in subproject KEP directories (e.g., [trainer/docs/proposals](https://github.com/kubeflow/trainer/tree/master/docs/proposals)). The full KEP lifecycle (provisional -> implementable -> implemented) and template are documented.

- [X] **Document overview of project architecture and software design that demonstrates viable cloud native use cases, as part of the project's documentation.**

Architecture documentation is detailed: the [GTR](https://github.com/kubeflow/community/blob/master/KUBEFLOW-GENERAL-TECHNICAL-REVIEW.md) covers design principles, architecture requirements, service dependencies, IAM design, API topology, and release process. The [kubeflow.org architecture page](https://www.kubeflow.org/docs/started/architecture/) includes an AI lifecycle diagram and component descriptions. The [Kubeflow Community Distribution release pages](https://www.kubeflow.org/docs/kubeflow-distribution/releases/) provide per-release component version matrices.

> - [ ] **Document the project's release process and guidelines publicly in a RELEASES.md or equivalent file that defines:**

Reviewer: ```suggestion
- [x] **Document the project's release process and guidelines publicly in a RELEASES.md or equivalent file that defines:**
```

# Kubeflow Graduation Due Diligence

- Link to [Graduation Application: Kubeflow](https://github.com/cncf/toc/issues/1861)
- TOC Sponsor: [@kfaseela](https://github.com/kfaseela)
- Review started: May 4, 2026
- Working notes: [kubeflow-graduation-dd-notes.md](./kubeflow-graduation-dd-notes.md)

## Graduation Evaluation Summary for Kubeflow

### Criteria Evaluation

[Faseela K](https://github.com/kfaseela) conducted the due diligence for Kubeflow, which applied for Graduation from Incubating status (accepted 2023-07-23).

The following criteria implementations are noteworthy:

> - [Kubeflow](https://www.kubeflow.org/) is a modular, Kubernetes-native AI/ML platform that provides a foundation of composable tools for building and operating ML workflows at scale. This graduation covers **6 subprojects**: Kubeflow Trainer, Kubeflow Pipelines, Kubeflow Katib, Kubeflow Notebooks, Kubeflow Spark Operator, and Kubeflow Model Registry (repo: kubeflow/hub). Each subproject satisfies the graduation criteria independently.

Reviewer 2: @kfaseela Do you want to modify the description as:
```suggestion
- [Kubeflow](https://www.kubeflow.org/) is a modular, Kubernetes-native AI/ML platform that provides a foundation of composable tools for building and operating AI/ML workloads at scale. This graduation covers **6 subprojects**: Kubeflow Trainer, Kubeflow Pipelines, Kubeflow Katib, Kubeflow Notebooks, Kubeflow Spark Operator, and Kubeflow Model Registry (repo: kubeflow/hub). Each subproject satisfies the graduation criteria independently.
```

We believe that ML workflows are very broad term, and might confuse users that Kubeflow is alternative to other workflow tools like Airflow, Flyte, or Metaflow, which is not correct.

- Kubeflow's [GitOps-based access control](https://github.com/kubeflow/internal-acls) via Peribolos and Prow is particularly well-designed: org membership, team membership, and repo permissions are declared in a versioned YAML file and reconciled against GitHub, with 2FA enforced for all org members.

> - The project's governance is multi-layered, with a Kubeflow Steering Committee (KSC, 5 seats, max 1 per organization) for strategic decisions and Working Groups owning each subproject. The 2026 KSC election completed on schedule, with all seats filled and no organization holding more than 1 seat.

Reviewer 2: Do you want to mention anything about Kubeflow Outreach Committee and, recently created, Kubeflow Distribution Committee?

- A [General Technical Review (GTR)](https://github.com/kubeflow/community/blob/master/KUBEFLOW-GENERAL-TECHNICAL-REVIEW.md) was completed in September 2025, providing comprehensive architecture, design, and engineering documentation that satisfies several DD criteria. GTR snapshot is also submitted to cncf/toc via [Add Kubeflow GTR snapshot (toc#2180)](https://github.com/cncf/toc/pull/2180).
- All 6 graduation subprojects hold at least an OpenSSF Best Practices Passing badge; Kubeflow Trainer and Kubeflow Katib have achieved Gold.
- The Kubeflow Ecosystem page and process were established during this review in response to DD findings, clarifying the boundary between core subprojects and ecosystem partners (KServe, Feast, Elyra). KServe, previously presented under the Kubeflow brand, is now clearly identified as an independent CNCF Incubating project.
- The Kubeflow maintainers and community have been exceptionally responsive throughout this review - PRs were merged, issues filed and assigned, and maintainers engaged constructively on all feedback.

The following blocking items were identified during the review and must be resolved before a graduation vote:

> - **[kubeflow/community#992](https://github.com/kubeflow/community/issues/992) - Track and resolve third-party security audit findings**: The Ada Logics audit report (Sept 2025) has been published at [security/Ada_Logics-Kubeflow-security-audit-2025.pdf](https://github.com/kubeflow/community/blob/master/security/Ada_Logics-Kubeflow-security-audit-2025.pdf) via [Publish Ada Logics security audit (community#994)](https://github.com/kubeflow/community/pull/994) - good progress. However, 12 of 14 findings (3 Critical, 6 Moderate, 2 Low) have no public tracking or fix links. The graduation criterion requires moderate and low findings to be planned/tracked for resolution. Maintainers must link where each Critical and Moderate finding is fixed or tracked (issue/PR/advisory).

Reviewer 2: You should be able to find PR links here: https://github.com/kubeflow/community/issues/992#issuecomment-4782945952

- [X] **Document how the project makes decisions on leadership, contribution acceptance, requests to the CNCF, and changes to governance or project goals.**

Decision-making is documented across three processes: (1) KSC decisions ([committee-steering/charter.md](https://github.com/kubeflow/community/blob/master/committee-steering/charter.md)) - Normal Decisions (quorum + majority) and Special Decisions (70% threshold) for governance/charter changes; (2) [KEP process](https://github.com/kubeflow/community/tree/master/proposals) for user/operator-facing or major architectural changes; (3) [Ecosystem join process](https://github.com/kubeflow/community/tree/master/ecosystem) for external project partnerships (KSC vote required).

- [X] **Document how role, function-based members, or sub-teams are assigned, onboarded, and removed for specific teams (example: Security Response Committee).**

> The [contributor ladder](https://www.kubeflow.org/docs/about/membership/) documents concrete criteria for Member, Reviewer, and Approver roles. WG lifecycle (creation, retirement) and WG Chair selection are documented in the refactored [wg-governance.md](https://github.com/kubeflow/community/blob/master/committee-steering/wg-governance.md): Chairs must be at least a community Member, hold a documented subproject role, have sustained contributions, be sponsored by a member, and be confirmed by super-majority vote of existing Chairs. Responsibilities/privileges and inactive Chair removal are also explicitly documented. All private Google Doc references removed. Resolved via [Refactor WG governance and Chair requirements (community#989)](https://github.com/kubeflow/community/pull/989).

Reviewer 2: ```suggestion
The [contributor ladder](https://github.com/kubeflow/community/blob/master/community-membership.md) documents concrete criteria for Member, Reviewer, and Approver roles. WG lifecycle (creation, retirement) and WG Chair selection are documented in the refactored [wg-governance.md](https://github.com/kubeflow/community/blob/master/committee-steering/wg-governance.md): Chairs must be at least a community Member, hold a documented subproject role, have sustained contributions, be sponsored by a member, and be confirmed by super-majority vote of existing Chairs. Responsibilities/privileges and inactive Chair removal are also explicitly documented. All private Google Doc references removed. Resolved via [Refactor WG governance and Chair requirements (community#989)](https://github.com/kubeflow/community/pull/989).
```

- [X] **Document complete list of current maintainers, including names, contact information, domain of responsibility, and affiliation.**

[MAINTAINERS.md](https://github.com/kubeflow/community/blob/master/MAINTAINERS.md) links to the KSC member list, KOC member list, WG chairs/leads in `wgs.yaml`, and per-subproject OWNERS files. The KSC and KOC lists include names, GitHub handles, organizations, and terms. Staleness in `wgs.yaml` is being addressed by [Remove outdated WGs/SIGs (community#974)](https://github.com/kubeflow/community/pull/974).

- [X] **A number of active maintainers which is appropriate to the size and scope of the project.**

> The [CNCF project-maintainers.csv](https://github.com/cncf/foundation/blob/main/project-maintainers.csv) lists 45 Kubeflow maintainers. Per OWNERS file approver counts across the 6 graduation subprojects: Trainer (8), Pipelines (6), Katib (4), Spark Operator (7), Model Registry (8), Notebooks (2) - 35 approvers total. Maintainer count is appropriate for a project of Kubeflow's size.

Reviewer: > Per OWNERS file approver counts across the 6 graduation subprojects: [...] Notebooks (2)

Notebooks has 4 active approvers on it's development branches:

`main` & `notebooks-v1` (given that https://github.com/kubeflow/notebooks/pull/1214 (which cleans up inactive approvers) is being merged):
```
$ find . -name "OWNERS" | xargs yq ".approvers[]" | grep -v -- "---" | sort --uniq
andyatmiami
thesuperzapper
```

`notebooks-v2`:
```
$ find . -name "OWNERS" | xargs yq ".approvers[]" | grep -v -- "---" | sort --uniq
andyatmiami
ederign
paulovmr
thesuperzapper
```

cc @andyatmiami @thesuperzapper

- [X] **CNCF Code of Conduct is cross-linked from other governance documents.**

The CNCF CoC is referenced in GOVERNANCE.md, wg-governance.md, and individual WG charters. As of 2026-06-18, standalone `CODE_OF_CONDUCT.md` files are present in all 7 graduation subproject repos: community, spark-operator, trainer, katib, notebooks, hub, and pipelines. Resolved via [kubeflow/community#976](https://github.com/kubeflow/community/issues/976).

- [X] **All subprojects, if any, are listed.**

The 6 graduation subprojects are listed consistently across the graduation application, GTR, kubeflow.org architecture page, `wgs.yaml`, and `internal-acls/org.yaml`:
1. Kubeflow Trainer - [kubeflow/trainer](https://github.com/kubeflow/trainer)
2. Kubeflow Pipelines - [kubeflow/pipelines](https://github.com/kubeflow/pipelines)
3. Kubeflow Katib - [kubeflow/katib](https://github.com/kubeflow/katib)
4. Kubeflow Notebooks - [kubeflow/notebooks](https://github.com/kubeflow/notebooks)
5. Kubeflow Spark Operator - [kubeflow/spark-operator](https://github.com/kubeflow/spark-operator)

> 6. Kubeflow Model Registry - [kubeflow/hub](https://github.com/kubeflow/hub)

Reviewer 3: ```suggestion
6. Kubeflow Hub - [kubeflow/hub](https://github.com/kubeflow/hub)
```

We renamed Kubeflow Model Registry to Kubeflow Hub recently. We probably want to replace throughout:
```
sed -i 's/Model Registry/Hub/g'
```
